File: src/main/models/upload_seen.py
from django.db import models
from main.models import AppUser

class UploadSeen(models.Model):
    upload = models.ForeignKey("UploadAction", on_delete=models.CASCADE)
    user = models.ForeignKey(AppUser, on_delete=models.CASCADE)
    seen_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        unique_together = ("upload", "user")
        indexes = [
            models.Index(fields=["user", "seen_at"]),
        ]

File: backend/src/main/java/fr/speciarium/model/UploadSeen.java
package fr.speciarium.model;

import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
import jakarta.persistence.*;

import java.time.OffsetDateTime;

@Entity
@Table(
    name = "main_uploadseen",
    uniqueConstraints = @UniqueConstraint(columnNames = {"upload_id", "user_id"}),
    indexes = @Index(columnList = "user_id, seen_at")
)
public class UploadSeen extends PanacheEntityBase {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "upload_id")
    public UploadAction upload;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id")
    public AppUser user;

    @Column(name = "seen_at", nullable = false, updatable = false)
    public OffsetDateTime seenAt;

    @PrePersist
    void prePersist() {
        if (seenAt == null) seenAt = OffsetDateTime.now();
    }
}
